File: mod/26.1.x/neoforge/src/main/java/net/wiredtomato/waygl/neoforge/NeoForgeWayGL.java
package net.wiredtomato.waygl.neoforge;

import net.minecraft.client.gui.screens.TitleScreen;
import net.minecraft.network.chat.Component;
import net.neoforged.api.distmarker.Dist;
import net.neoforged.bus.api.IEventBus;
import net.neoforged.fml.ModContainer;
import net.neoforged.fml.common.Mod;
import net.neoforged.fml.loading.FMLConfig;
import net.neoforged.neoforge.client.event.ScreenEvent;
import net.neoforged.neoforge.client.gui.IConfigScreenFactory;
import net.neoforged.neoforge.common.NeoForge;
import net.wiredtomato.waygl.WayGL;
import net.wiredtomato.waygl.config.ConfigManager;
import net.wiredtomato.waygl.core.Loader;
import net.wiredtomato.waygl.core.os.OS;
import net.wiredtomato.waygl.neoforge.screen.WarningScreen;

@Mod(value = WayGL.MOD_ID, dist = Dist.CLIENT)
public class NeoForgeWayGL {
    private boolean gaveWarning = false;

    public NeoForgeWayGL(IEventBus modBus, ModContainer container) {
        WayGL.init();
        NeoForge.EVENT_BUS.addListener(this::onScreenOpen);

        container.registerExtensionPoint(IConfigScreenFactory.class, (configContainer, modListScreen) ->
                ConfigManager.createConfigScreen(modListScreen)
        );
    }

    private void onScreenOpen(ScreenEvent.Opening event) {
        if (gaveWarning || OS.current() != OS.LINUX || !(event.getScreen() instanceof TitleScreen) || Loader.isWayland()) return;

        var earlyWindowControl = FMLConfig.getBoolConfigValue(FMLConfig.ConfigValue.EARLY_WINDOW_CONTROL);
        if (earlyWindowControl) {
            event.setNewScreen(new WarningScreen(Component.translatable("waygl.warning.early_window_active"), event.getScreen()));
        }

        gaveWarning = true;
    }
}

File: mod/neoforge/src/main/kotlin/net/wiredtomato/waygl/NeoForgeWayGL.kt
package net.wiredtomato.waygl

import net.minecraft.client.gui.screens.TitleScreen
import net.minecraft.network.chat.Component
import net.neoforged.api.distmarker.Dist
import net.neoforged.fml.common.Mod
import net.neoforged.fml.loading.FMLConfig
import net.neoforged.neoforge.client.event.ScreenEvent
import net.neoforged.neoforge.common.NeoForge.EVENT_BUS
import net.wiredtomato.waygl.client.screen.WarningScreen
import thedarkcolour.kotlinforforge.neoforge.forge.runWhenOn

@Mod(WayGL.MODID)
object NeoForgeWayGL {
    private var gaveWarning = false

    init {
        WayGL.tryUseWayland()

        runWhenOn(Dist.CLIENT) {
            EVENT_BUS.addListener(::onScreenOpen)
        }
    }

    private fun onScreenOpen(event: ScreenEvent.Opening) {
        if (gaveWarning || event.screen !is TitleScreen || WayGL.isWayland) return

        val earlyWindowControl = FMLConfig.getBoolConfigValue(FMLConfig.ConfigValue.EARLY_WINDOW_CONTROL)
        if (earlyWindowControl) {
            event.newScreen = (WarningScreen(Component.translatable("waygl.screen.warning"), event.screen))
        }

        gaveWarning = true
    }
}